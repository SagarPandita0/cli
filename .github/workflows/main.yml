name: CI

on: [push, pull_request]

jobs:
  build:
    strategy:
      matrix:
        platform: [ubuntu-latest]
        node: ['14', '16']
    name: Node ${{ matrix.node }} (${{ matrix.platform }})
    runs-on: ${{ matrix.platform }}
    steps:
      - uses: actions/checkout@v1
      - uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node }}
      - name: install dependencies
        run: yarn install --frozen-lockfile --ignore-engines
      - name: lint code
        run: npm run lint
      - name: run tests
        run: npm run test
      - name: get code coverage report
        run: npx codecov
        env:
          CODECOV_TOKEN: ${{secrets.CODECOV_TOKEN}}

  release:
    name: do semantic release
    runs-on: 'ubuntu-latest'
    environment: publish
    needs: build
    steps:
      - uses: actions/checkout@v1
      - uses: actions/setup-node@v1
        with:
          node-version: '16'
      - name: install dependencies
        run: yarn install --frozen-lockfile --ignore-engines
      - name: release
        run: npx run semantic-release
        env:
          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
          NPM_TOKEN: ${{secrets.NPM_TOKEN}}
        if: github.event_name == 'push' && github.ref == 'refs/heads/main'
