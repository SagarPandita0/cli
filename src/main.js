import path from 'path'

import { getAllFiles } from './utils/fs.js'
import { convertToJson } from './utils/yaml-parser.js'
import { getEventsMd, formatToMarkdown } from './utils/md-formatter.js'
import { createYearBuckets, getEventsStats } from './utils/content-manager.js'

export { formatToMarkdown, generateGigs, generateDocument }

async function generateDocument({ sourceDirectory, preContent, postContent }) {
  let markdownOutputPreContent = ''
  let markdownOutputPostContent = ''
  let document = ''

  const eventsMarkdown = await generateGigs({ sourceDirectory })

  if (preContent) {
    markdownOutputPreContent = processCustomContent(preContent)
  }

  if (postContent) {
    markdownOutputPostContent = processCustomContent(postContent)
  }

  document +=
    markdownOutputPreContent + eventsMarkdown + markdownOutputPostContent

  return document
}

function processCustomContent(contents) {
  let markdownContent = ''
  for (const content of contents) {
    if (content.hasOwnProperty('raw')) {
      markdownContent += content.raw + '\n'
    }

    if (content.hasOwnProperty('format')) {
      markdownContent += formatToMarkdown(content.format) + '\n'
    }
  }

  return markdownContent
}

async function generateGigs({ sourceDirectory }) {
  let directoryPath = path.join(process.cwd(), sourceDirectory)
  if (path.isAbsolute(sourceDirectory)) {
    directoryPath = sourceDirectory
  }

  const allFilesFlatList = await getAllFiles(directoryPath)

  const entries = []
  for (const filePath of allFilesFlatList) {
    const json = await convertToJson(filePath)
    entries.push(json)
  }

  const entriesByBucket = await getEntriesByBuckets(entries)
  const entriesForYearMarkdown = getEventsMd(entriesByBucket.bucketsByYear)
  return entriesForYearMarkdown
}

async function getEntriesByBuckets(entries) {
  // @TODO these buckets:
  //  const bucketsAll = []
  //  const bucketsType = {}
  //  const bucketsLanguage = {}
  //  const bucketsCountry = {}

  const bucketsByYear = createYearBuckets(entries)
  const eventsStats = getEventsStats(entries)

  return {
    bucketsByYear,
    eventsStats
  }
}
