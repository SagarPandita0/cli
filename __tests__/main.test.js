import { URL } from 'url'
import path from 'path'
import { generateDocument } from '../src/main.js'

const __dirname = new URL('.', import.meta.url).pathname

test('given a directory of markdown files a full document is rendered', async () => {
  const filePath = path.join(__dirname, './__fixtures__/main-datafiles')
  const jsonResult = await generateDocument({ sourceDirectory: filePath })
  expect(jsonResult).toMatchSnapshot()
})

test('given a directory of markdown files a full document is rendered along with pre and post content', async () => {
  const filePath = path.join(__dirname, './__fixtures__/main-datafiles')
  const jsonResult = await generateDocument({
    sourceDirectory: filePath,
    preContent: [
      {
        raw: '<p> actual html is allowed too </p>'
      },
      {
        format: [
          {
            p: 'Liran Tal'
          }
        ]
      }
    ],
    postContent: [
      {
        raw: '<p align="center"> rendered in postcontent </p>'
      },
      {
        format: [
          {
            p: 'Thank you!'
          }
        ]
      }
    ]
  })
  expect(jsonResult).toMatchSnapshot()
})
