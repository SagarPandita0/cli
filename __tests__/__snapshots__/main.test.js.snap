// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a directory of markdown files a full document is rendered 1`] = `
"# Table of Contents



"
`;

exports[`given a directory of markdown files a full document is rendered along with pre and post content 1`] = `
"<p> actual html is allowed too </p>

Liran Tal

# Table of Contents



<p align=\\"center\\"> rendered in postcontent </p>

Thank you!

"
`;
